### ISSUE TITLE
Add percentile pause info as well as avg pause

### ISSUE THREAD

**Issue by elyograg** [chewiebug/GCViewer#139]
The average pause is useful information, but the median pause (50th percentile) would be a lot more useful.  Also helpful would be 75th, 95th, and 99th percentile information.

I think the info required to calculate these values is already mostly available, in order to draw the graphs.  I don't know how much additional memory would be required.  For a very large GC log, it could be quite a lot.

https://en.wikipedia.org/wiki/Percentile

**Comment by ecki** [comment]
I agree (especially if we get percentiles over periods). But its quite some work. 

I had a quick hack back then (not sure if it still applies) which allows me to specify a threshold of STW pause times which I care about, and get a count for all those "outliers". Thats not a percentile but similiar usefull for SLA monitoring. https://github.com/ecki/GCViewer/commit/7ca9236a8d056e038c813627678b5732898dc511

**Comment by elyograg** [comment]
If you want to avoid huge amounts of additional memory use, you could do it in the same way that codahale metrics does.  Or you could simply incorporate codahale metrics directly into your application, but it would increase the jar size.  The Solr project has included codahale metrics to give users percentile information on query times in the admin UI.

https://issues.apache.org/jira/browse/SOLR-1972

From what I understand, codahale builds a 1024-element construct to store the info being tracked (don't know if it's an array or a Collection) ... when it fills up, they randomly pick one of the existing entries to evict and use that smaller data set to calculate the percentiles.  It doesn't give you *exactly* the same information as you would get from a complete dataset, but apparently it is pretty close.

**Comment by duckdeer** [comment]
This would be a very nice enhancement, too. I think it can be a big help analyzing big logfiles. 

Percentiles would be great, too. :-) 

Am 9. Juli 2015 00:54:50 MESZ, schrieb Bernd <notifications@github.com>:
>I agree (especially if we get percentiles over periods). But its quite
>some work. 
>
>I had a quick hack back then (not sure if it still applies) which
>allows me to specify a threshold of STW pause times which I care about,
>and get a count for all those "outliers". Thats not a percentile but
>similiar usefull for SLA monitoring.
>https://github.com/ecki/GCViewer/commit/7ca9236a8d056e038c813627678b5732898dc511
>
>---
>Reply to this email directly or view it on GitHub:
>https://github.com/chewiebug/GCViewer/issues/139#issuecomment-119753591

-- 
Diese Nachricht wurde von meinem Android-Mobiltelefon mit K-9 Mail gesendet.

**Comment by chewiebug** [comment]
Hmm, this looks like I should start thinking on how to implement the percentiles :-).

**Comment by chewiebug** [comment]
I have made a first quick and dirty implementation of the percentile calculation. The values are shown on the "event details" tab.

It is pushed to the [feature/#139/percentile](https://github.com/chewiebug/GCViewer/tree/feature/%23139/percentile) branch. 

What do you think?

**Comment by elyograg** [comment]
Now I just need to figure out how to build the project!

**Comment by duckdeer** [comment]
You must build the project with maven and java 8. 
Just go to the main directory and type "mvn clean install". After that you'll find the jar file under "target". 


Am 15. Juli 2015 17:31:06 MESZ, schrieb Shawn Heisey <notifications@github.com>:
>Now I just need to figure out how to build the project!
>
>---
>Reply to this email directly or view it on GitHub:
>https://github.com/chewiebug/GCViewer/issues/139#issuecomment-121653114

-- 
Diese Nachricht wurde von meinem Android-Mobiltelefon mit K-9 Mail gesendet.

**Comment by elyograg** [comment]
I figured out how to build it, and looked at the Event details tab.  Looks pretty awesome!

One number that I didn't see is something that might be called "total pause" ... which would only be something to worry about if there are situations where a GC pause includes multiple pause types.  If that's not a situation that can ever occur, then I guess it's not a number that needs to be gathered.

Nitpick, feel free to ignore:  Some people might be specifically looking for "Median" rather than the 50th percentile.

**Comment by elyograg** [comment]
Thought of something else.  The "Pause" tab has min/avg/max for "Pause interval" ... would it be possible to have statistics in "Event details" for that number?
